RESTEasy Reactive fails to handle collections of parameterized types as parameter

### Describe the bug

We try to switch from RESTEasy Classic to RESTEasy Reactive and checked the differences in the [migration guide](https://quarkus.io/guides/resteasy-reactive-migration).

The problem is now, that (with RESTEasy Reactive) we get an exception during the startup, because we have methods defined, which uses a List/Set of parameterized types, like in the following example:

```
@GET
@Path("/wrapperList")
public String wrapperList(@QueryParam("wrapperList") final List<WrapperClass<StatusEnum>> wrapperList) {
    // ...
}
```

For the `WrapperClass` we defined a `ParameterConverter`, which worked fine with RESTEasy Classic and also when we provide a single parameter to a method (not a collection) in RESTEasy Reactive.

Example (dummy implementation):

```
public class WrapperClassParamConverter implements ParamConverter<WrapperClass<?>> {

    private final Class<?> rawType;
    private final Type genericType;

    public WrapperClassParamConverter(
            final Class<?> rawType,
            final Type genericType
    ) {
        this.rawType = rawType;
        this.genericType = genericType;
    }

    @Override
    public WrapperClass<?> fromString(String value) {
        return new WrapperClass<>(Enum.valueOf(StatusEnum.class, value));
    }

    @Override
    public String toString(WrapperClass<?> wrapperClass) {
        return wrapperClass != null ? wrapperClass.getValue().toString() : null;
    }

}
```


The exception (complete stacktrace can be found in "Actual behavior"):

```
Caused by: java.lang.IllegalArgumentException: Unable to support parameter converter with type: 'org.acme.WrapperClass<org.acme.StatusEnum>'
        at org.jboss.resteasy.reactive.server.core.startup.RuntimeResourceDeployment.smartInitParameterConverter(RuntimeResourceDeployment.java:544)
        at org.jboss.resteasy.reactive.server.core.startup.RuntimeResourceDeployment.buildResourceMethod(RuntimeResourceDeployment.java:366)
        at org.jboss.resteasy.reactive.server.core.startup.RuntimeDeploymentManager.deploy(RuntimeDeploymentManager.java:136)
        at io.quarkus.resteasy.reactive.server.runtime.ResteasyReactiveRecorder.createDeployment(ResteasyReactiveRecorder.java:135)
        at io.quarkus.deployment.steps.ResteasyReactiveProcessor$setupDeployment713137389.deploy_0(Unknown Source)
        at io.quarkus.deployment.steps.ResteasyReactiveProcessor$setupDeployment713137389.deploy(Unknown Source)
        at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
...
```


The code worked fine with RESTEasy Classic, but fails with RESTEasy Reactive. Because of this and as there is no hint in the migration guide, I declare it as a bug.

When we have to write additional parameter converters, please let me know. I failed to find a working solution.

### Expected behavior

Parameter converters should be used like in RESTEasy Classic - so also for collections.

### Actual behavior

Exception is thrown

```
2023-09-06 11:59:34,539 ERROR [io.qua.run.boo.StartupActionImpl] (Quarkus Main Thread) Error running Quarkus: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:104)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.ExceptionInInitializerError
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:70)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:124)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
        ... 6 more
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
        ... 15 more
Caused by: java.lang.IllegalArgumentException: Unable to support parameter converter with type: 'org.acme.WrapperClass<org.acme.StatusEnum>'
        at org.jboss.resteasy.reactive.server.core.startup.RuntimeResourceDeployment.smartInitParameterConverter(RuntimeResourceDeployment.java:544)
        at org.jboss.resteasy.reactive.server.core.startup.RuntimeResourceDeployment.buildResourceMethod(RuntimeResourceDeployment.java:366)
        at org.jboss.resteasy.reactive.server.core.startup.RuntimeDeploymentManager.deploy(RuntimeDeploymentManager.java:136)
        at io.quarkus.resteasy.reactive.server.runtime.ResteasyReactiveRecorder.createDeployment(ResteasyReactiveRecorder.java:135)
        at io.quarkus.deployment.steps.ResteasyReactiveProcessor$setupDeployment713137389.deploy_0(Unknown Source)
        at io.quarkus.deployment.steps.ResteasyReactiveProcessor$setupDeployment713137389.deploy(Unknown Source)
        ... 16 more

```

### How to Reproduce?

Reproducer: https://github.com/timonzi/resteasy-reactive-wrapper-collection

1. Start the application using `mvn compile quarkus:dev` --> exception is thrown
2. Change the Maven dependency from `quarkus-resteasy-reactive-jackson` to `quarkus-resteasy-jackson` --> no exception
3. Validate that with RESTEasy Classic the mentioned method works fine using: GET http://localhost:8080/parameters/wrapperList?wrapperList=ACTIVE&wrapperList=INACTIVE
4. When switching back to `quarkus-resteasy-reactive-jackson`, you can comment out or remove the `wrapperList` method - then the startup works

### Output of `uname -a` or `ver`

Linux nb-timonz 6.2.0-31-generic #31~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Aug 16 13:45:26 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

OpenJDK Runtime Environment Temurin-17.0.6+10 (build 17.0.6+10)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.3.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.1 (2e178502fcdbffc201671fb2537d0cb4b4cc58f8)

### Additional information

_No response_

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/simple/GenericsParamConverterTest.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/simple/GenericsParamConverterTest.java
@@ -1,0 +1,109 @@
+package io.quarkus.resteasy.reactive.server.test.simple;
+
+import static io.restassured.RestAssured.given;
+import static org.hamcrest.CoreMatchers.is;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.QueryParam;
+import jakarta.ws.rs.ext.ParamConverter;
+import jakarta.ws.rs.ext.ParamConverterProvider;
+import jakarta.ws.rs.ext.Provider;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class GenericsParamConverterTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest test = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(TestEnum.class, Wrapper.class,
+                            WrapperParamConverterProvider.class, WrapperParamConverterProvider.WrapperParamConverter.class,
+                            TestResource.class));
+
+    @Test
+    public void wrapper() {
+        given()
+                .when().get("/test/single?wrapper=ACTIVE")
+                .then()
+                .statusCode(200)
+                .body(is("ACTIVE"));
+    }
+
+    @Test
+    public void wrapperList() {
+        given()
+                .when().get("/test/list?wrapperList=INACTIVE&wrapperList=ACTIVE")
+                .then()
+                .statusCode(200)
+                .body(is("INACTIVE,ACTIVE"));
+    }
+
+    @Path("/test")
+    public static class TestResource {
+
+        @GET
+        @Path("/list")
+        public String list(@QueryParam("wrapperList") final List<Wrapper<TestEnum>> wrapperList) {
+            return wrapperList.stream().map(w -> w.getValue().name()).collect(Collectors.joining(","));
+        }
+
+        @GET
+        @Path("/single")
+        public String single(@QueryParam("wrapper") final Wrapper<TestEnum> wrapper) {
+            return wrapper.getValue().toString();
+        }
+    }
+
+    public enum TestEnum {
+        ACTIVE,
+        INACTIVE
+    }
+
+    public static class Wrapper<E extends Enum<E>> {
+        private final E value;
+
+        public Wrapper(final E value) {
+            this.value = value;
+        }
+
+        public E getValue() {
+            return value;
+        }
+    }
+
+    @Provider
+    public static class WrapperParamConverterProvider implements ParamConverterProvider {
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public <T> ParamConverter<T> getConverter(final Class<T> rawType, final Type genericType,
+                final Annotation[] annotations) {
+            if (Wrapper.class.isAssignableFrom(rawType)) {
+                return (ParamConverter<T>) new WrapperParamConverter();
+            }
+            return null;
+        }
+
+        public static class WrapperParamConverter implements ParamConverter<Wrapper<?>> {
+
+            @Override
+            public Wrapper<?> fromString(String value) {
+                return new Wrapper<>(Enum.valueOf(TestEnum.class, value));
+            }
+
+            @Override
+            public String toString(Wrapper<?> wrapper) {
+                return wrapper != null ? wrapper.getValue().toString() : null;
+            }
+        }
+    }
+}

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/startup/RuntimeResourceDeployment.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/startup/RuntimeResourceDeployment.java
@@ -529,6 +529,8 @@ public class RuntimeResourceDeployment {
                     Type genericType = genericArguments[0];
                     if (genericType instanceof Class) {
                         genericTypeClassName = ((Class<?>) genericType).getName();
+                    } else if (genericType instanceof ParameterizedType) {
+                        genericTypeClassName = ((ParameterizedType) genericType).getRawType().getTypeName();
                     } else if (genericType instanceof WildcardType) {
                         WildcardType genericTypeWildcardType = (WildcardType) genericType;
                         Type[] upperBounds = genericTypeWildcardType.getUpperBounds();

